File: Week5/course.py
# -*- coding: utf-8 -*-
"""
@author: AEROJASH
"""

import numpy as np

# Constants
MAX_STUDENTS = 30

# Global variables to store course data
num_students = 0
names = []
surnames = []
dnis = []
grades = np.zeros((MAX_STUDENTS, 10), dtype=float)

def add_student(name, surname, dni, student_grades):
    """
    Add a new student to the course.

    Parameters:
    - name (str): The first name of the student.
    - surname (str): The last name of the student.
    - dni (str): The DNI (National Identity Document) of the student.
    - student_grades (list of float): List of grades for the student (up to 10 grades).

    Returns:
    - None
    """
    global num_students, names, surnames, dnis, grades
    if num_students < MAX_STUDENTS:
        names.append(name)
        surnames.append(surname)
        dnis.append(dni)
        grades[num_students, :] = student_grades
        num_students += 1
        print(f"Student {name} {surname} added successfully!")
    else:
        print("Maximum number of students reached!")

def get_course_list():
    """
    Get a list of all students in the course.

    Returns:
    - list of str: List containing the names and surnames of all students in the course.
    """
    return [f"{surnames[i]}, {names[i]}" for i in range(num_students)]

def calculate_course_average():
    """
    Calculate the average grade of the entire course.

    Returns:
    - float: The average grade of the course.
    """
    return np.mean(grades[:num_students])

def calculate_student_average(student_index):
    """
    Calculate the average grade of a specific student.

    Parameters:
    - student_index (int): Index of the student in the course.

    Returns:
    - float: The average grade of the student.
    """
    return np.mean(grades[student_index, :])

def get_student_with_highest_grade():
    """
    Get the name of the student with the highest overall grade in the course.

    Returns:
    - str: The name and surname of the student with the highest grade.
    """
    max_grade_index = np.argmax(np.sum(grades[:num_students], axis=1))
    return f"{names[max_grade_index]} {surnames[max_grade_index]}"

def get_student_with_lowest_grade():
    """
    Get the name of the student with the lowest overall grade in the course.

    Returns:
    - str: The name and surname of the student with the lowest grade.
    """
    min_grade_index = np.argmin(np.sum(grades[:num_students], axis=1))
    return f"{names[min_grade_index]} {surnames[min_grade_index]}"

# Example usage:
add_student("John", "Doe", "123456789", [80, 85, 90, 75, 95, 88, 92, 87, 84, 89])
add_student("Jane", "Smith", "987654321", [70, 75, 80, 65, 85, 78, 82, 77, 74, 79])

print("Course List:", get_course_list())
print("Course Average Grade:", calculate_course_average())
print("Average Grade for John Doe:", calculate_student_average(0))
print("Student with Highest Grade:", get_student_with_highest_grade())
print("Student with Lowest Grade:", get_student_with_lowest_grade())

File: Week6/Course/src/coursePackage/Course.java
package coursePackage;
import java.util.ArrayList;

public class Course {

    /**
     * Course class
     * @param MAX_STUDENTS: Maximum number of students
     * @param num_students: Number of students
     * @param students: List of students
     */

    private int MAX_STUDENTS = 30;
    private int num_students = 0;
    private ArrayList<Student> students;

    public Course() {

        /**
         * Constructor
         * @param students: List of students
         */

        this.students = new ArrayList<Student>();
    }

    public void addStudent(Student student) {
        
        /**
         * Adds a student to the course
         * @param student: Student object
         * @return void
         * @exception Maximum number of students reached
         */

        if (num_students < MAX_STUDENTS) {
            students.add(student);
            num_students++;
            System.out.println("Student "+ student.getName() + " " + student.getSurname() + " added successfully!");
        } else {
            System.out.println("Maximum number of students reached");
        }
    }

    public ArrayList<String> getStudents() {

        /**
         * Returns a list with the students' names and surnames
         * @return ArrayList<String>
         */

        ArrayList<String> studentsInfo = new ArrayList<String>();
        for (Student student : students) {
            studentsInfo.add(student.getName() + " " + student.getSurname());
        }
        return studentsInfo;
    }

    public Double calculate_course_average(){

        /**
         * Calculates the average grade of all students
         * @return Double
         */

        Double total = 0.0;
        int total_grades = 0;
        for (Student student : students) {
            for (Double grade : student.getGrades()) {
                total += grade;
                total_grades++;
            }
        }
        return total/total_grades;
    }

    public Double calculate_student_average(int index){

        /**
         * Calculates the average grade of a student
         * @param index: Index of the student in the list
         * @return Double
         */

        Double total = 0.0;
        for (Double grade : students.get(index).getGrades()) {
            total += grade;
        }
        return total/students.get(index).getGrades().size();
    }

    public String get_student_with_highest_grade(){

        /**
         * Returns the student with the highest grade
         * @return String
         */

        int index = 0;
        Double highest_grade = 0.0;
        for(int i = 0; i < students.size(); i++){
            if (calculate_student_average(i) >= highest_grade){
                highest_grade = calculate_student_average(i);
                index = i;
            }
        }
        return students.get(index).getName() + " " + students.get(index).getSurname();
    }

    public String get_student_with_lowest_grade(){

        /**
         * Returns the student with the lowest grade
         * @return String
         */

        int index = 0;
        Double lowest_grade = 100.0;
        for(int i = 0; i < students.size(); i++){
            if (calculate_student_average(i) <= lowest_grade){
                lowest_grade = calculate_student_average(i);
                index = i;
            }
        }
        return students.get(index).getName() + " " + students.get(index).getSurname();
    }
}
